tools/lib/git-tools.bash

```diff
@@ -0,0 +1,34 @@
+# shellcheck shell=bash
+
+# Borrowed from Git's git-sh-setup.
+#
+# See git.git commit 92c62a3f4 (from 2010!); as of 2020 with Git 2.26,
+# this function has only needed one edit since then, adding localization
+# with gettext, which we can omit.
+require_clean_work_tree () {
+    local action="$1"
+
+    git rev-parse --verify HEAD >/dev/null || exit 1
+    git update-index -q --ignore-submodules --refresh
+    local err=0
+
+    if ! git diff-files --quiet --ignore-submodules; then
+        echo >&2 "Cannot $action: You have unstaged changes."
+        err=1
+    fi
+
+    if ! git diff-index --cached --quiet --ignore-submodules HEAD --; then
+        if [ $err = 0 ]; then
+            echo >&2 "Cannot $action: Your index contains uncommitted changes."
+        else
+            echo >&2 "Additionally, your index contains uncommitted changes."
+        fi
+        err=1
+    fi
+
+    if [ $err = 1 ]; then
+        git status --short
+        echo >&2 "Doing nothing to avoid losing your work."
```

Comment:
In `./tools/review` I check for a pristine checkout with `git status --porcelain`.  Is that technique too naive, or is it applicable here?

~~~
 29 def check_git_pristine():
 30     # type: () -> None
 31     command = ['git', 'status', '--porcelain']
 32     output = check_output(command)
 33     if output.strip():
 34         exit('Git is not pristine:\n' + output)

Reply:
Interesting comparison!

I think that likely works. I'd expect it to do the equivalent of `update-index --refresh`, so avoiding the main bug that motivated this PR. And certainly it'll catch the edge case where the index has changes but the worktree agrees with HEAD.

The output is a bit less nice than `git status --short` -- in particular the latter uses colors, the same as regular `git status` does, if the user has `color.ui=auto`.

There are a couple of other small differences in behavior which are unlikely to show up in the Zulip context:
* `require_clean_work_tree` checks `git rev-parse --verify HEAD` up front. The common case I know of where this would fail is if you've just done `git init` and haven't made any commits.
* `require_clean_work_tree` passes `--ignore-submodules` to the various commands. I'm not totally convinced that's actually the right behavior, in cases where it makes a difference. It probably speeds things up, if you have big submodules. In any case we don't use submodules in Zulip and I don't expect we ever will, because despite some good ideas the implementation makes them a bad developer experience.

So I guess the relevant differences in behavior are:
* that color difference
* I'm not sure enough of my understanding of the subsystem that `git update-index --refresh` is operating on to be more than like 95% confident that the `git status --porcelain` solution doesn't have the bug that fixes.